src/theory/bags/normal_form.cpp

```diff
@@ -12,26 +12,628 @@
+    elements[itA->first] = itA->second;
+  };
+
+  auto greaterOrEqual = [](std::map<Node, Rational>& elements,
+                           std::map<Node, Rational>::const_iterator& itA,
+                           std::map<Node, Rational>::const_iterator& itB) {
+    // itB->first is not in A, so we just skip it
+  };
+
+  auto remainderOfA = [](std::map<Node, Rational>& elements,
+                         std::map<Node, Rational>& elementsA,
+                         std::map<Node, Rational>::const_iterator& itA) {
+    // append the remainder of A
+    while (itA != elementsA.end())
+    {
+      elements[itA->first] = itA->second;
+      itA++;
+    }
+  };
+
+  auto remainderOfB = [](std::map<Node, Rational>& elements,
+                         std::map<Node, Rational>& elementsB,
+                         std::map<Node, Rational>::const_iterator& itB) {
+    // do nothing
+  };
+
+  return evaluateBinaryOperation(
+      n, equal, less, greaterOrEqual, remainderOfA, remainderOfB);
+}
+
+Node NormalForm::evaluateChoose(TNode n)
+{
+  Assert(n.getKind() == BAG_CHOOSE);
+  // Examples
+  // --------
+  // - (choose (emptyBag String)) = "" // the empty string which is the first
+  //   element returned by the type enumerator
+  // - (choose (MK_BAG "x" 4)) = "x"
+  // - (choose (union_disjoint (MK_BAG "x" 4) (MK_BAG "y" 1))) = "x"
+  //     deterministically return the first element
+
+  if (n[0].getKind() == EMPTYBAG)
+  {
+    TypeNode elementType = n[0].getType().getBagElementType();
+    TypeEnumerator typeEnumerator(elementType);
+    // get the first value from the typeEnumerator
+    Node element = *typeEnumerator;
+    return element;
+  }
+
+  if (n[0].getKind() == MK_BAG)
+  {
+    return n[0][0];
+  }
+  Assert(n[0].getKind() == UNION_DISJOINT);
+  // return the first element
+  // e.g. (choose (union_disjoint (MK_BAG "x" 4) (MK_BAG "y" 1)))
+  return n[0][0][0];
+}
+
+Node NormalForm::evaluateCard(TNode n)
+{
+  Assert(n.getKind() == BAG_CARD);
+  // Examples
+  // --------
+  //  - (card (emptyBag String)) = 0
+  //  - (choose (MK_BAG "x" 4)) = 4
+  //  - (choose (union_disjoint (MK_BAG "x" 4) (MK_BAG "y" 1))) = 5
+
+  std::map<Node, Rational> elements = getBagElements(n[0]);
+  Rational sum(0);
+  for (std::pair<Node, Rational> element : elements)
+  {
+    sum += element.second;
+  }
+
+  NodeManager* nm = NodeManager::currentNM();
+  Node sumNode = nm->mkConst(sum);
+  return sumNode;
+}
+
+Node NormalForm::evaluateIsSingleton(TNode n)
+{
+  Assert(n.getKind() == BAG_IS_SINGLETON);
+  // Examples
+  // --------
+  // - (bag.is_singleton (emptyBag String)) = false
+  // - (bag.is_singleton (MK_BAG "x" 1)) = true
+  // - (bag.is_singleton (MK_BAG "x" 4)) = false
+  // - (bag.is_singleton (union_disjoint (MK_BAG "x" 1) (MK_BAG "y" 1))) = false
+
+  NodeManager* nm = NodeManager::currentNM();
+  Node falseNode = nm->mkConst(false);
+  Node trueNode = nm->mkConst(true);
+  if (n[0].getKind() == EMPTYBAG)
```

Comment:
> Can remove these 4 lines.

What do you mean?

Reply:
you don't need to check emptybag, this should just be part of the default "return false".

Reply:
Done